File: 37-sudoku-solver/sudoku-solver.java
class Solution {
    public void solveSudoku(char[][] board) {
        solve(board);
    }

    private boolean solve(char[][] board) {
        // Traverse all cells
        for (int row = 0; row < 9; row++) {
            for (int col = 0; col < 9; col++) {
                // If empty cell found
                if (board[row][col] == '.') {
                    // Try digits from 1 to 9
                    for (char num = '1'; num <= '9'; num++) {
                        if (isValid(board, row, col, num)) {
                            board[row][col] = num; // Place the number

                            // Recurse
                            if (solve(board)) return true;

                            // Backtrack if not valid
                            board[row][col] = '.';
                        }
                    }
                    return false; // No valid number found, must backtrack
                }
            }
        }
        return true; // Puzzle is solved
    }

    private boolean isValid(char[][] board, int row, int col, char num) {
        // Check row and column
        for (int i = 0; i < 9; i++) {
            if (board[row][i] == num || board[i][col] == num)
                return false;
        }

        // Check 3x3 grid
        int startRow = 3 * (row / 3);
        int startCol = 3 * (col / 3);

        for (int r = startRow; r < startRow + 3; r++) {
            for (int c = startCol; c < startCol + 3; c++) {
                if (board[r][c] == num)
                    return false;
            }
        }

        return true;
    }
}

File: 37-sudoku-solver/sudoku-solver.cpp
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        solve(board);
    }
    
private:
    bool solve(vector<vector<char>>& board) {
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                if (board[r][c] == '.') {
                    for (char d = '1'; d <= '9'; d++) {
                        if (isValid(board, r, c, d)) {
                            board[r][c] = d;
                            if (solve(board)) return true; // solved
                            board[r][c] = '.'; // backtrack
                        }
                    }
                    return false; // no valid digit found
                }
            }
        }
        return true; // all cells filled
    }
    
    bool isValid(vector<vector<char>>& board, int row, int col, char val) {
        for (int i = 0; i < 9; i++) {
            // check row
            if (board[row][i] == val) return false;
            // check column
            if (board[i][col] == val) return false;
            // check 3x3 box
            int boxRow = (row / 3) * 3 + i / 3;
            int boxCol = (col / 3) * 3 + i % 3;
            if (board[boxRow][boxCol] == val) return false;
        }
        return true;
    }
};
